

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>prot_db &mdash; IntegronFinder 2.0.6 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=3e466467"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="results" href="results.html" />
    <link rel="prev" title="integron" href="integron.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            IntegronFinder
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../user_guide/index.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/installation.html">Installation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/installation.html#integronfinder-dependencies">IntegronFinder dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/installation.html#installation-procedure">Installation procedure</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/installation.html#from-version-2-0">From Version 2.0</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/installation.html#from-version-1-5-1-and-after">From Version 1.5.1 and after</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/installation.html#for-version-1-5-and-before">For Version 1.5 and before</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/installation.html#uninstallation-procedure">Uninstallation procedure</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/installation.html#id2">From Version 1.5.1 and after</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/installation.html#from-version-1-0-to-version-1-5">From Version 1.0 to Version 1.5</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/installation.html#how-to-install-python">How to install Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/changes.html">What’s new ?</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/changes.html#v2-0-6">V2.0.6</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/changes.html#v2-0-5">V2.0.5</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/changes.html#v2-0-3">V2.0.3</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/changes.html#in-version-2-0">In Version 2.0</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/tutorial.html">Quick start</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/tutorial.html#for-impatient">For impatient</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/tutorial.html#input-and-outputs">Input and Outputs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#inputs">Inputs</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#custom-protein-file">Custom protein file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#outputs">Outputs</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/tutorial.html#for-everyone">For everyone</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#thorough-local-detection">Thorough local detection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#calin-detection">CALIN detection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#functional-annotation">Functional annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#search-for-promoter-and-atti-sites">Search for promoter and <em>attI</em> sites</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#keep-intermediate-results">Keep intermediate results</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#topology">Topology</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/tutorial.html#for-big-data-people">For big data people</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#parallelization">Parallelization</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/tutorial.html#for-integron-diggers">For integron diggers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#clustering-of-elements">Clustering of elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#integrase">Integrase</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#attc-evalue"><em>attC</em> evalue</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#attc-size"><em>attC</em> size</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#palindromes">Palindromes</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/tutorial.html#attc-alignements"><em>attC</em> alignements</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/webserver.html">web server</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../user_guide/webserver.html#galaxy">Galaxy</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/webserver.html#how-to-use-it">How to use it</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../user_guide/webserver.html#results">Results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../user_guide/references.html">References</a></li>
</ul>
</li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Developer Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../dev_guide.html">Developer installation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../dev_guide.html#dependencies">Dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../dev_guide.html#integron-finder-testing-procedures">Integron_Finder testing procedures</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../dev_guide.html#coverage">Coverage</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../dev_guide.html#documentation">Documentation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../dev_guide.html#build-a-new-release">Build a new release</a></li>
<li class="toctree-l3"><a class="reference internal" href="../dev_guide.html#send-changes-to-upstream-repository">Send changes to upstream repository</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../overview.html">Architecture Overview</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../overview.html#project-files-and-directories">Project files and directories</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../overview.html#files">Files</a></li>
<li class="toctree-l4"><a class="reference internal" href="../overview.html#directories">Directories</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../overview.html#technical-overview">Technical overview</a></li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Integron_finder API Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="annotation.html">annotation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="annotation.html#integron_finder.annotation.add_feature"><code class="docutils literal notranslate"><span class="pre">add_feature()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="annotation.html#integron_finder.annotation.func_annot"><code class="docutils literal notranslate"><span class="pre">func_annot()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="attc.html">attc</a><ul>
<li class="toctree-l4"><a class="reference internal" href="attc.html#integron_finder.attc.find_attc_max"><code class="docutils literal notranslate"><span class="pre">find_attc_max()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="attc.html#integron_finder.attc.search_attc"><code class="docutils literal notranslate"><span class="pre">search_attc()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="config.html">config</a><ul>
<li class="toctree-l4"><a class="reference internal" href="config.html#integron_finder.config.Config"><code class="docutils literal notranslate"><span class="pre">Config</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hmm.html">hmm</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hmm.html#integron_finder.hmm.read_hmm"><code class="docutils literal notranslate"><span class="pre">read_hmm()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hmm.html#integron_finder.hmm.scan_hmm_bank"><code class="docutils literal notranslate"><span class="pre">scan_hmm_bank()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="infernal.html">infernal</a><ul>
<li class="toctree-l4"><a class="reference internal" href="infernal.html#integron_finder.infernal.expand"><code class="docutils literal notranslate"><span class="pre">expand()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="infernal.html#integron_finder.infernal.find_attc"><code class="docutils literal notranslate"><span class="pre">find_attc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="infernal.html#integron_finder.infernal.local_max"><code class="docutils literal notranslate"><span class="pre">local_max()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="infernal.html#integron_finder.infernal.read_infernal"><code class="docutils literal notranslate"><span class="pre">read_infernal()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="integrase.html">integrase</a><ul>
<li class="toctree-l4"><a class="reference internal" href="integrase.html#integron_finder.integrase.find_integrase"><code class="docutils literal notranslate"><span class="pre">find_integrase()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="integron.html">integron</a><ul>
<li class="toctree-l4"><a class="reference internal" href="integron.html#integron_finder.integron.Integron"><code class="docutils literal notranslate"><span class="pre">Integron</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="integron.html#integron_finder.integron.find_integron"><code class="docutils literal notranslate"><span class="pre">find_integron()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">prot_db</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#integron_finder.prot_db.GembaseDB"><code class="docutils literal notranslate"><span class="pre">GembaseDB</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#integron_finder.prot_db.ProdigalDB"><code class="docutils literal notranslate"><span class="pre">ProdigalDB</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#integron_finder.prot_db.ProteinDB"><code class="docutils literal notranslate"><span class="pre">ProteinDB</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#integron_finder.prot_db.SeqDesc"><code class="docutils literal notranslate"><span class="pre">SeqDesc</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="results.html">results</a><ul>
<li class="toctree-l4"><a class="reference internal" href="results.html#integron_finder.results._log"><code class="docutils literal notranslate"><span class="pre">_log</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="results.html#integron_finder.results.filter_calin"><code class="docutils literal notranslate"><span class="pre">filter_calin()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="results.html#integron_finder.results.integrons_report"><code class="docutils literal notranslate"><span class="pre">integrons_report()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="results.html#integron_finder.results.merge_results"><code class="docutils literal notranslate"><span class="pre">merge_results()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="results.html#integron_finder.results.summary"><code class="docutils literal notranslate"><span class="pre">summary()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="topology.html">topology</a><ul>
<li class="toctree-l4"><a class="reference internal" href="topology.html#integron_finder.topology.Topology"><code class="docutils literal notranslate"><span class="pre">Topology</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="utils.html">utils</a><ul>
<li class="toctree-l4"><a class="reference internal" href="utils.html#integron_finder.utils.FastaIterator"><code class="docutils literal notranslate"><span class="pre">FastaIterator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="utils.html#integron_finder.utils.MultiFastaReader"><code class="docutils literal notranslate"><span class="pre">MultiFastaReader</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="utils.html#integron_finder.utils.get_name_from_path"><code class="docutils literal notranslate"><span class="pre">get_name_from_path()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="utils.html#integron_finder.utils.log_level"><code class="docutils literal notranslate"><span class="pre">log_level()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="utils.html#integron_finder.utils.model_len"><code class="docutils literal notranslate"><span class="pre">model_len()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">IntegronFinder</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Developer Guide</a></li>
          <li class="breadcrumb-item"><a href="index.html">Integron_finder API Reference</a></li>
      <li class="breadcrumb-item active">prot_db</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/developer_guide/api/prot_db.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="prot-db">
<span id="id1"></span><h1>prot_db<a class="headerlink" href="#prot-db" title="Link to this heading"></a></h1>
<p>The prot_db module contains classes to handle protein file and protein description
which can be either generate by Prodigal or Provide by Gembase.
It also provide an interface to abstract the way to get protein sequences and descriptions</p>
<dl class="py class" id="module-integron_finder.prot_db">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">integron_finder.prot_db.</span></span><span class="sig-name descname"><span class="pre">GembaseDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gembase_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB" title="Link to this definition"></a></dt>
<dd><p>Implements <a class="reference internal" href="#integron_finder.prot_db.ProteinDB" title="integron_finder.prot_db.ProteinDB"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProteinDB</span></code></a> from a Gembase.
Managed proteins from Proteins directory corresponding to a replicon/contig</p>
<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_seq_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.__getitem__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>prot_seq_id</strong> (<em>str</em>) – the id of a protein sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Sequence corresponding to the prot_seq_id.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gembase_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.__init__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>replicon</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> object with a extra attribute path) – The replicon used to create ProteinDB (protein files and extra information)</p></li>
<li><p><strong>cfg</strong> (<a class="reference internal" href="config.html#integron_finder.config.Config" title="integron_finder.config.Config"><code class="xref py py-class docutils literal notranslate"><span class="pre">integron_finder.config.Config</span></code></a> object) – The integron_finder configuration</p></li>
<li><p><strong>prot_file</strong> – The path to a protein file in fasta format
which is the translation of the replicon</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The replicon is a modified Bio.SeqRecord object.
The attribute <em>path</em> must be injected in the object
This attribute represent the path to a fasta file representing this replicon</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.__iter__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on the gene seq_ids which constitute the contig (genes and pseudogenes).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generator</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB._make_protfile">
<span class="sig-name descname"><span class="pre">_make_protfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB._make_protfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB._make_protfile" title="Link to this definition"></a></dt>
<dd><p>Create fasta file with protein corresponding to this sequence, from the corresponding Gembase protfile
This step is necessary because in Gembase1&amp;2 Draft
One nucleic file can contain several contigs, but all proteins are in the same file.
and in Gembase2 replicons file can contain several chromosomes and plasmids
:return: the path of the created protein file
:rtype: str</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB._parse_lst">
<span class="sig-name descname"><span class="pre">_parse_lst</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB._parse_lst"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB._parse_lst" title="Link to this definition"></a></dt>
<dd><p>Parse the LSTINFO file and extract information specific to the replicon
:return: <cite>class</cite>:pandas.DataFrame` object
:raise IntegronError: when LST dir is not found</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.coding_prot_ids">
<span class="sig-name descname"><span class="pre">coding_prot_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.coding_prot_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.coding_prot_ids" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on coding genes seq_ids (the non coding genes are discarded)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generator</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.find_gembase_file_basename">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">find_gembase_file_basename</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gembase_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_seq_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.find_gembase_file_basename"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.find_gembase_file_basename" title="Link to this definition"></a></dt>
<dd><p>from the input file name, try to retrieve the basename which is used in gembase
This specially useful when IF is run in parallel. The input sequence is split
in chunks and treat in parallel. But in this case the name of the chunk does not match
neither the lstinfo file nor the protein file.
So this method try retrieve the original basename without extension
for instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ACBA</span><span class="mf">.0917.00019</span><span class="o">.</span><span class="n">fna</span>               <span class="o">=&gt;</span> <span class="n">ACBA</span><span class="mf">.0917.00019</span>
<span class="n">ACBA</span><span class="mf">.0917.00019.0001</span><span class="o">.</span><span class="n">fna</span>          <span class="o">=&gt;</span> <span class="n">ACBA</span><span class="mf">.0917.00019</span>
<span class="n">ESCO001</span><span class="o">.</span><span class="n">C</span><span class="mf">.00001</span><span class="o">.</span><span class="n">C001</span><span class="o">.</span><span class="n">fst</span>          <span class="o">=&gt;</span> <span class="n">ESCO001</span><span class="o">.</span><span class="n">C</span><span class="mf">.00001</span><span class="o">.</span><span class="n">C001</span>
<span class="n">ESCO001</span><span class="o">.</span><span class="n">C</span><span class="mf">.00001</span><span class="o">.</span><span class="n">C001_chunk_1</span><span class="o">.</span><span class="n">fst</span>  <span class="o">=&gt;</span> <span class="n">ESCO001</span><span class="o">.</span><span class="n">C</span><span class="mf">.00001</span><span class="o">.</span><span class="n">C001</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the gembase basename corresponding to the input file</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.gembase1_complete_parser">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gembase1_complete_parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.gembase1_complete_parser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.gembase1_complete_parser" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst_path</strong> (<em>str</em>) – the path of the LSTINFO file Gembase Complet</p></li>
<li><p><strong>sequence_id</strong> (<em>str</em>) – the id of the genomic sequence to analyse</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the information related to the ‘valid’ CDS corresponding to the sequence_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>class</cite>:pandas.DataFrame` object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.gembase1_draft_parser">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gembase1_draft_parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicon_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.gembase1_draft_parser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.gembase1_draft_parser" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst_path</strong> (<em>str</em>) – the path of the LSTINFO file from a Gembase Draft</p></li>
<li><p><strong>sequence_id</strong> (<em>str</em>) – the id of the genomic sequence to analyse</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the information related to the ‘valid’ CDS corresponding to the sequence_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>class</cite>:pandas.DataFrame` object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.gembase2_parser">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gembase2_parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicon_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.gembase2_parser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.gembase2_parser" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst_path</strong> (<em>str</em>) – the path of the LSTINFO file from a Gembase Draft</p></li>
<li><p><strong>sequence_id</strong> (<em>str</em>) – the id of the genomic sequence to analyse</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the information related to the ‘valid’ CDS corresponding to the sequence_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>class</cite>:pandas.DataFrame` object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.gembase_sniffer">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gembase_sniffer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.gembase_sniffer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.gembase_sniffer" title="Link to this definition"></a></dt>
<dd><p>Detect the type of gembase
:param str lst_path: the path to the LSTINFO file corresponding to the nucleic sequence
:returns: either the type of replicon (‘Complet’ or ‘Draft’, 1 or 2)
:rtype: tuple</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.get_description">
<span class="sig-name descname"><span class="pre">get_description</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.get_description"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.get_description" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_id</strong> (<em>str</em>) – a protein/gene identifier</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The description of the protein corresponding to the gene_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#integron_finder.prot_db.SeqDesc" title="integron_finder.prot_db.SeqDesc"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqDesc</span></code></a> namedtuple object</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IntegronError</strong> – when gene_id is not a valid Gembase gene identifier</p></li>
<li><p><strong>KeyError</strong> – if gene_id is not found in GembaseDB instance</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.get_lst_dir">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_lst_dir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gembase_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.get_lst_dir"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.get_lst_dir" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the Gembase LST directory</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.GembaseDB.get_replicon_type">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_replicon_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rep_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#GembaseDB.get_replicon_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.GembaseDB.get_replicon_type" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq_id</strong> (<em>str</em>) – the sequence id to parse</p></li>
<li><p><strong>rep_id</strong> (<em>str</em>) – the replicon identifier</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>the kind of genome, it can be either:</p>
<ul class="simple">
<li><p>Chromosome</p></li>
<li><p>Plasmid</p></li>
<li><p>Phage</p></li>
<li><p>Other</p></li>
<li><p>Draft</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RepliconType</span></code> object</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">integron_finder.prot_db.</span></span><span class="sig-name descname"><span class="pre">ProdigalDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB" title="Link to this definition"></a></dt>
<dd><p>Creates proteins from Replicon/contig using prodigal and provide facilities to access them.</p>
<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_seq_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB.__getitem__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>prot_seq_id</strong> (<em>str</em>) – the id of a protein sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Sequence corresponding to the prot_seq_id.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB.__iter__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on the protein seq_id which constitute the contig.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generator</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB._make_protfile">
<span class="sig-name descname"><span class="pre">_make_protfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB._make_protfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB._make_protfile" title="Link to this definition"></a></dt>
<dd><p>Use <cite>prodigal</cite> to generate proteins corresponding to the replicon</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the path of the created protfile</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB.coding_prot_ids">
<span class="sig-name descname"><span class="pre">coding_prot_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB.coding_prot_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB.coding_prot_ids" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on coding genes seq_ids (the non coding genes are discarded)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generartor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProdigalDB.get_description">
<span class="sig-name descname"><span class="pre">get_description</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProdigalDB.get_description"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProdigalDB.get_description" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_id</strong> (<em>str</em>) – a protein/gene identifier</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The description of the protein corresponding to the gene_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#integron_finder.prot_db.SeqDesc" title="integron_finder.prot_db.SeqDesc"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqDesc</span></code></a> namedtuple object</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IntegronError</strong> – when gene_id is not a valid Gembase gene identifier</p></li>
<li><p><strong>KeyError</strong> – if gene_id is not found in ProdigalDB instance</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">integron_finder.prot_db.</span></span><span class="sig-name descname"><span class="pre">ProteinDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB" title="Link to this definition"></a></dt>
<dd><p>AbstractClass defining the interface for ProteinDB.
ProteinDB provide an abstraction and a way to access to proteins corresponding
to the replicon/contig CDS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.__getitem__">
<em class="property"><span class="k"><span class="pre">abstractmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_seq_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.__getitem__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>prot_seq_id</strong> (<em>str</em>) – the id of a protein sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Sequence corresponding to the prot_seq_id.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> object</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>KeyError</strong> – when seq_id does not match any sequence in DB</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.__init__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>replicon</strong></p></li>
<li><p><strong>cfg</strong></p></li>
<li><p><strong>prot_file</strong></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.__iter__">
<em class="property"><span class="k"><span class="pre">abstractmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.__iter__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on the protein seq_id which constitute the contig.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generator</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.__weakref__">
<span class="sig-name descname"><span class="pre">__weakref__</span></span><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.__weakref__" title="Link to this definition"></a></dt>
<dd><p>list of weak references to the object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB._make_db">
<span class="sig-name descname"><span class="pre">_make_db</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB._make_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB._make_db" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>an index of the sequence contains in protfile corresponding to the replicon</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB._make_protfile">
<em class="property"><span class="k"><span class="pre">abstractmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_make_protfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB._make_protfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB._make_protfile" title="Link to this definition"></a></dt>
<dd><p>Create fasta file with protein corresponding to the nucleic sequence (replicon)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>the path of the created protein file</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.coding_prot_ids">
<em class="property"><span class="k"><span class="pre">abstractmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">coding_prot_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.coding_prot_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.coding_prot_ids" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a generator which iterate on coding genes seq_ids (the non coding genes are discarded)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>generartor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.get_description">
<em class="property"><span class="k"><span class="pre">abstractmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_description</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.get_description"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.get_description" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_id</strong> (<em>str</em>) – a protein/gene identifier</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The description of the protein corresponding to the gene_id</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#integron_finder.prot_db.SeqDesc" title="integron_finder.prot_db.SeqDesc"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqDesc</span></code></a> namedtuple object</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IntegronError</strong> – when gene_id is not a valid Gembase gene identifier</p></li>
<li><p><strong>KeyError</strong> – if gene_id is not found in GembaseDB instance</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.is_pseudo_gene">
<span class="sig-name descname"><span class="pre">is_pseudo_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/integron_finder/prot_db.html#ProteinDB.is_pseudo_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.is_pseudo_gene" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq_id</strong> – The sequence id to test</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the seq_id correspond to gene which is a pseudo gene, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="integron_finder.prot_db.ProteinDB.protfile">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">protfile</span></span><a class="headerlink" href="#integron_finder.prot_db.ProteinDB.protfile" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The absolute path to the protein file corresponding to contig id</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">integron_finder.prot_db.</span></span><span class="sig-name descname"><span class="pre">SeqDesc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.__getnewargs__">
<span class="sig-name descname"><span class="pre">__getnewargs__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.__getnewargs__" title="Link to this definition"></a></dt>
<dd><p>Return self as a plain tuple.  Used by copy and pickle.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.__new__">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__new__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.__new__" title="Link to this definition"></a></dt>
<dd><p>Create new instance of SeqDesc(id, strand, start, stop)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.__replace__">
<span class="sig-name descname"><span class="pre">__replace__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.__replace__" title="Link to this definition"></a></dt>
<dd><p>Return a new SeqDesc object replacing specified fields with new values</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return a nicely formatted representation string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc._asdict">
<span class="sig-name descname"><span class="pre">_asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc._asdict" title="Link to this definition"></a></dt>
<dd><p>Return a new dict which maps field names to their values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc._make">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_make</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">iterable</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc._make" title="Link to this definition"></a></dt>
<dd><p>Make a new SeqDesc object from a sequence or iterable</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc._replace">
<span class="sig-name descname"><span class="pre">_replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc._replace" title="Link to this definition"></a></dt>
<dd><p>Return a new SeqDesc object replacing specified fields with new values</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.id" title="Link to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.start">
<span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.start" title="Link to this definition"></a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.stop">
<span class="sig-name descname"><span class="pre">stop</span></span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.stop" title="Link to this definition"></a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="integron_finder.prot_db.SeqDesc.strand">
<span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#integron_finder.prot_db.SeqDesc.strand" title="Link to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="integron.html" class="btn btn-neutral float-left" title="integron" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="results.html" class="btn btn-neutral float-right" title="results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015-2025 Institut Pasteur, Paris, CNRS..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/Integron_Finder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>