
Welcome.

SingleM is a tool for profiling shotgun metagenomes. It has a particular strength in detecting microbial lineages which are not in reference databases. The method it uses also makes it suitable for some related tasks, such as assessing eukaryotic contamination, finding bias in genome recovery, and lineage-targeted MAG recovery.

Documentation can be found at https://wwood.github.io/singlem/

Community profiles for many publicly available metagenomes can be found at the [Sandpiper](https://sandpiper.qut.edu.au/) website.

### Citations
The overall citation for SingleM/Sandpiper:

Woodcroft, Ben J., Samuel TN Aroney, Rossen Zhao, Mitchell Cunningham, Joshua AM Mitchell, Linda Blackall, and Gene W. Tyson. SingleM and Sandpiper: Robust microbial taxonomic profiles from metagenomic data. bioRxiv (2024): 2024-01. https://doi.org/10.1101/2024.01.30.578060

The `microbial_fraction` (SMF) mode of SingleM:

Eisenhofer, Raphael, Antton Alberdi, and Ben J. Woodcroft. Large-scale estimation of bacterial and archaeal DNA prevalence in metagenomes reveals biome-specific patterns. bioRxiv (2024): 2024-05. https://doi.org/10.1101/2024.05.16.594470
